### ISSUE TITLE
Typing passphrases only once with chezmoi+age

### ISSUE THREAD

**Issue by spinscale** [twpayne/chezmoi#2005]
## What exactly are you trying to do?

I have been trying to use chezmoi with age (binary installed) in order to encrypt certain files. This is working great. I tested out the symmetic encryption and symmetric encryption with a passphrase as mentioned in the [docs](https://www.chezmoi.io/user-guide/encryption/age/). One thing I have not been able to figure out: Is it possible to type a passphrase only once, even when several/all files are encrypted with the same passphrase? So far nothing I tried has been working and I was forced to type the passphrase the number of times I have encrypted files when running something like `chezmoi status`. I mainly want to figure out, if this works at all with age and I just did a mistake or if I have to go with GPG instead.

Thanks a ton for helping, this is a great tool!

**Comment by twpayne** [comment]
chezmoi currently assumes that you're using age with asymmetric encryption, with the private key stored in plaintext on your local machine, as using a passphrase would mean that you'd have to enter your password every time you run chezmoi (age, unlike GnuPG, does not have an agent). As age supports multiple recipients, you can ensure that the local private key can only decrypt the secrets that you want on that machine. It would be possible to make chezmoi prompt and store the passphrase (as it does for KeepassXC) but this isn't implemented yet.

Personally, entering the passphrase every time I run chezmoi is too annoying for me (note that chezmoi will only prompt if actually needs to decrypt something). If you think you can tolerate this, I can add a feature as you describe so that you only get prompted once.

What do you think?

**Comment by spinscale** [comment]
Thanks a ton for answering so quick!

If the whole issue/problem does not exist, when running with the GPG agent enabled, I'd say there is no need for such a feature and everyone should just use GPG then. No need to do things twice.

**Comment by twpayne** [comment]
I dug a bit further into age's support for files encrypted with a passphrase:
* When using the `age` command, there is no way to pass the passphrase programmatically. `age` gets a terminal to prompt for the passphrase, and I don't think there's a way to override this.
* Even if chezmoi could pass the passphrase programmatically, there's no guarantee that all files will be encrypted with the same passphrase.
* It's possible to work around the above by using age as a library, but it's quite a lot of effort to implement a feature that I think few people would use.

A possible reasonable workflow would be:
* Generate an age private key. Encrypt your files with this and check the encrypted files into your dotfiles repo.
* Encrypt the private key with a passphrase, and then add that encrypted private key to your repo.
* Use a chezmoi `run_once_` script to decrypt the private key on new machines.
This would mean you'd only need to type your passphrase when you set up your dotfiles on a new machine.

In the meantime, I'll close this issue as `won't fix` as I think there are better alternatives.
